### PR TITLE
🧐(project) fetch inactive users from edx mysql database

### PR DESCRIPTION
**Pull request by wilbrdt**
## Purpose

To warn/delete users that have been inactive for a specified duration, integrate SQLAlchemy models to connect to the edX database and retrieve user data.

### CODE REVIEW

File: src/mork/edx/database.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                url, echo=settings.EDX_DB_DEBUG, pool_pre_ping=True
+            )
+        if session is not None:
+            self.session = session
+        else:
+            self.session = Session(self.engine)
+
+        self.User = User
+
+    def get_inactive_users(
```
Comment (lunika) [inline src/mork/edx/database.py:42]: I didn't find anything in the SQLAlchemy documentation about where should be declared this kind of method. I don't think doing it here is a good idea. The joanie import project was more a one shot project and we will delete it soon.
Here it is not the case and I think this kind of method should not be all declare in this class.

In an ORM like doctrine, a repository can be used and it contains all this kind of method related to a model

Comment (wilbrdt) [reply src/mork/edx/database.py]: You're right, it should not be in this class. As this particular query will be used in the two tasks (warning the users, delete the users), I think it should be nice to have a function for that.
What about having it in a `utils.py`?

Comment (wilbrdt) [reply src/mork/edx/database.py]: I pushed a proposal where I add a class method to the User model through a mixin. Let me know what you think of it!

File: src/mork/edx/mixins.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+
+class UserMixin:
+    """Mixin that adds convenience methods for the User operations."""
+
+    @classmethod
+    def get_inactives(
+        cls,
+        session: Session,
+        inactivity_period: timedelta,
```
Comment (quitterie-lcs) [inline src/mork/edx/mixins.py:17]: Do we set a default value here or when the method is called in tasks ?

File: tests/edx/test_mixins.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        3, last_login=Faker().date_time_between(end_date="-3y")
+    )
+    # 4 users that logged in recently
+    EdxUserFactory.create_batch(
+        4, last_login=Faker().date_time_between(start_date="-3y")
+    )
+
+    # Get all users inactive for more than 3 years
+    users = User.get_inactives(
+        db.session, inactivity_period=timedelta(days=365 * 3), offset=0, limit=9
```
Comment (quitterie-lcs) [inline tests/edx/test_mixins.py:24]: Why do you set limit to 9 ?

File: tests/edx/test_mixins.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    users = User.get_inactives(
+        db.session, inactivity_period=timedelta(days=365 * 3), offset=0, limit=9
+    )
+
+    assert users == []
+
+
+def test_edx_usermixin_get_inactives_slice(db):
+    """Test the `get_inactives` method with a slice."""
```
Comment (quitterie-lcs) [inline tests/edx/test_mixins.py:42]:
```suggestion
    """Test the `get_inactives` method with a slice from limit parameters."""
```

File: tests/edx/test_mixins.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    users = User.get_inactives(
+        db.session, inactivity_period=timedelta(days=365 * 3), offset=0, limit=2
+    )
+
+    assert len(users) == 2
+    assert users == inactive_users[:2]
+
+
+def test_edx_usermixin_get_inactives_slice_empty(db):
+    """Test the `get_inactives` method with an empty slice ."""
```
Comment (quitterie-lcs) [inline tests/edx/test_mixins.py:62]:
```suggestion
    """Test the `get_inactives` method with an empty slice."""
```

File: tests/edx/test_mixins.py (hunk 3 above)
Comment (wilbrdt) [reply tests/edx/test_mixins.py:24]: Good point, changing them to `0`

Comment (wilbrdt) [reply tests/edx/test_mixins.py]: Reverting to `9`, as for SQLAlchemy and for SQL, a limit of `0` does not mean unlimited, but means returning an empty set

Comment (lunika) [reply tests/edx/test_mixins.py:24]: And so, why 9 ? :sweat_smile:

**Review by lunika — APPROVED** [review]

**Pull request merged**